Infeasibility analyzer : improve unit tests

guilpier-code opened the pull request:
Mainly add new unit tests for slack analysis.
For details, some **gitHub** comments were left.

guilpier-code committed:
Renaming infeasibility analyzer : renaming in class Constraint

guilpier-code committed:
Renaming infeasibility analyzer : renaming in class InfeasibleProblemReport

guilpier-code committed:
Simplifying infeasibility analyzer

guilpier-code committed:
New constraint to detect by infeasibility analyzer

flomnes committed:
format

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/simplifying-infeasibility-analyzer

guilpier-code committed:
Infeasibility - to more changeability : simplification + renaming

guilpier-code committed:
Infeasibility - to more changeability : simplifying function StringBetweenAngleBrackets

guilpier-code committed:
Infeasibility - to more changeability : simplifying functions areaName() and timeStep()

guilpier-code committed:
Infeasibility - to more changeability : renaming Constraint::getType() into type()

guilpier-code committed:
Infeasibility - to more changeability : function STSName() ==> simplified and renamed into STSname()

guilpier-code committed:
Infeasibility - to more changeability : remove unnecessary comments

guilpier-code committed:
Infeasibility - to more changeability : simplifying function sortConstraintsByType()

Check on the current constraint's name number of components is useless : it cannot be null.

guilpier-code committed:
Infeasibility - to more changeability : in unit tests, change misleading BC names

guilpier-code committed:
Infeasibility - to more changeability : split constraints logs into 2 parts (suspicious constraints and infeasibility causes)

guilpier-code committed:
Infeasibility - to more changeability : move call to sortConstraintsByType with his friends

guilpier-code committed:
Infeasibility - to more changeability : run clang-format

guilpier-code committed:
Make infeasibily more changeable : move sorting and trimming slack vars from report to analysis

guilpier-code committed:
Make infeasibily more changeable : introducing constraints list to be detected by slack analysis

guilpier-code committed:
Make infeasibily more changeable : use new list of constraint and remove old style

guilpier-code committed:
Make infeasibily more changeable : remove old class Constraint (now useless) + move a header to the right location

guilpier-code committed:
Make infeasibily more changeable : renaming - detected ==> watched (for variables, classes, files)

guilpier-code committed:
Make infeasibily more changeable : split function addSlackVariables into 2 functions

guilpier-code committed:
Make infeasibily more changeable : clang format

guilpier-code committed:
Make infeasibily more changeable : correction due to review

guilpier-code committed:
Make infeasibily more changeable : add use of a factory and remove the clone function

guilpier-code committed:
[skip ci] Make infeasibily more changeable : renaming

guilpier-code committed:
Make infeasibily more changeable : clang format

guilpier-code committed:
Make infeasibily more changeable : avoid useless data member + avoid passing constraint factory

guilpier-code committed:
[skip ci] Make infeasibily more changeable : correct function naming

guilpier-code committed:
Make infeasibility more changeable : simplify constraint factory

guilpier-code committed:
Make infeasibility more changeable : simplifying constraint classes

guilpier-code committed:
Make infeasibility more changeable : fix compilation error on Linux + clang format

guilpier-code committed:
[skip ci] Make infeasibility more changeable ; very small change

guilpier-code committed:
Make infeasibility more changeable : use a std algorithm instead of a for loop

guilpier-code committed:
[skip ci] Make infeasibility more changeable : renaming

guilpier-code committed:
Make infeasibility more changeable : simplify constraint classes yet again

guilpier-code committed:
Make infeasibility more changeable : replace a loop with the use of std::copy_if

guilpier-code committed:
[skip ci] Make infeasibility more changeable : remove useless include

guilpier-code committed:
Make infeasibility more changeable : use of ranges algorithm instead of classic std algorithm

guilpier-code committed:
[skip ci] Make infeasibility more changeable : clang format

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/infeasibility-more-changeable

guilpier-code committed:
Improve infeasible pb test : improve constraint violation detection

guilpier-code committed:
Improve infeasible pb test : fix detection error detection + change report messages printings

guilpier-code committed:
Improve infeasible pb test : fix bug in logging violated variables + add tests on report slack analysis

guilpier-code committed:
Improve infeasible pb test : test slack analysis results (list of slack variables, ordered and trimmed)

guilpier-code committed:
Make infeasibility more changeable : simplifying constraint factory

guilpier-code committed:
Make infeasibility more changeable : corrections due to review

flomnes committed:
Minor fixes, use std::unique_ptr

flomnes committed:
Caller safer

flomnes committed:
Revert

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/infeasibility-more-changeable

flomnes committed:
Fix Sonar issues on #2232 (#2257)

Fix Sonar issues on #2232

flomnes committed:
Minor fixes, use std::unique_ptr, add safety (#2256)

- Use `std::unique_ptr` since we don't share objects
- Remove lambda functions (see
https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique)
- Use `const std::string&`
- Make check on `create`, add safety on no result found
- Remove unused variables in tests

guilpier-code committed:
Improve infeasible pb test : make creation of a problem with infeasible constraints more generic

guilpier-code committed:
Improve infeasible pb test : adding a test on limit of 10 to detected numbers of constraints

guilpier-code committed:
Improve infeasible pb test : replace a simple loop on integers with a loop involving std::ranges::iota_view(...)

guilpier-code committed:
Improve infeasible pb test : fixing a compilation error on Linux (<format> not available)

guilpier-code committed:
Make infeasibility more changeable : changing inner class constant into a static const

guilpier-code committed:
Make infeasibility more changeable : trial to simplify method InfeasibleProblemReport :: filterConstraintsToOneByType()

guilpier-code committed:
Make infeasibility more changeable : fix filterConstraintsToOneByType (used to fail on unit tests)

guilpier-code committed:
Make infeasibility more changeable : simplify filterConstraintsToOneByType

guilpier-code committed:
Improve infeasible pb test : small improvements on unit tests

guilpier-code committed:
Make infeasibility more changeable : fix compilation error

guilpier-code committed:
Improve infeasible pb test : restore a wrong removal

guilpier-code committed:
Make infeasibility more changeable : add comments to make code more understandable

flomnes committed:
Use std::type_index (#2264)

flomnes committed:
Formatting

flomnes committed:
Fix linking

guilpier-code committed:
Improve infeasible pb test : add a new test on reporting infeasibility causes + preliminary work

guilpier-code committed:
Make infeasibility more changeable : fix an error in infeasibility causes order in he report

guilpier-code committed:
Improve infeasible pb test : removing a useless test

guilpier-code committed:
Format

guilpier-code commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -50,7 +50,7 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
     if (slackVariables_.empty())
     {
-        logs.error() << title() << " : no constraints have been selected";
+        logs.warning() << title() << " : no constraints have been selected";
```
If no constraint was selected for analysis, do we have to raise an error ? Or a warning's enough ?

flomnes replied:
Yes, good enough

guilpier-code commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -63,10 +63,12 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
-
     sortSlackVariablesByValue();
     trimSlackVariables();
+    if (hasDetectedInfeasibilityCause_ = anySlackVariableNonZero(); !hasDetectedInfeasibilityCause_)
```
New behavior : print a message when no constraint violation is detected

guilpier-code commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -130,11 +132,24 @@ void ConstraintSlackAnalysis::trimSlackVariables()
-    report.logInfeasibilityCauses();
+    report.storeSuspiciousConstraints();
+    report.storeInfeasibilityCauses();
+    std::ranges::for_each(report.getLogs(), [](auto& line) { logs.notice() << line; });
```
For test purpose, the report object now produces a vector of logs (can be retrieved with method **getLogs()**).
And indeed, logging is not supposed to be done by a report builder (here an instance of **InfeasibleProblemReport**).

guilpier-code commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -36,7 +36,7 @@ std::string StringBetweenAngleBrackets(const std::string& constraintName)
 std::string timeStep(std::vector<std::string> splitName)
 {
-    return StringBetweenAngleBrackets(splitName.at(splitName.size() - 2));
+    return StringBetweenAngleBrackets(splitName.rbegin()[1]);
```
We could even call `splitName.end()[-2]` instead.

guilpier-code commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -203,16 +238,109 @@ BOOST_AUTO_TEST_CASE(analysis_should_ignore_ill_named_constraint)
+    BOOST_CHECK(!analysis.hasDetectedInfeasibilityCause());
+}
+
+BOOST_AUTO_TEST_CASE(analysis_slack_variables_are_ordered)
```
Here some new unit tests were added : 
- Unit tests for slack analysis class (the result of this class [at end of method **run()**]  is a collection of slack variable, ordered by slack value).
  - check that slack variables collection is ordered in expected order
  - check that slack variables collection is limited to 10 elements
- Unit tests for the report builder
  - all violated incoming constraints are reported
  - infeasibility causes are unique and sorted by slack value

guilpier-code committed:
[skip ci] Improve infeasible pb test : very small fix

guilpier-code commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -24,8 +24,6 @@
-
 namespace Antares::Optimization
 {
 InfeasibleProblemReport::InfeasibleProblemReport(
```
class **InfeasibleProblemReport** : 
Actually, what it does is : build a report for slack variable analysis.
So a better name could be : **SlackAnalysisReportBuilder**.
The associated file name could change accordingly.

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/improve-infeasible-pb-tests

flomnes commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h:
```diff
@@ -52,15 +52,19 @@ class ConstraintSlackAnalysis: public UnfeasibilityAnalysis
         return "Slack variables analysis";
     }
 
+    std::vector<const operations_research::MPVariable*>& largestSlackVariables();
```
```suggestion
    const std::vector<const operations_research::MPVariable*>& largestSlackVariables();
```

guilpier-code replied:
Done

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -21,13 +21,16 @@
 #define WIN32_LEAN_AND_MEAN
 #define BOOST_TEST_MODULE unfeasible_problem_analyzer
 
+#include <cstdio>
```
```suggestion
```

guilpier-code replied:
Done

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
+    std::unique_ptr<MPSolver> problem(MPSolver::CreateSolver("GLOP"));
+    for (auto i: std::ranges::iota_view(1, n + 1)) // From 1 to n included
+    {
+        char name[32];
+        std::sprintf(name, "BC-name-%d::hourly::hour<%d>", i, 5 * i);
```
Use `std::ostringstream` or `std::to_string` ? `sprintf` is C

guilpier-code replied:
Yes. Originally I used **std::format** on **Win**, unfortunately not available on Linux.

guilpier-code replied:
Done by using **std::stringstream**

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
+    auto* var = problem->MakeNumVar(varLowBnd, varUpBnd, varName);
+    auto* constraint = problem->MakeRowConstraint(name);
+    constraint->SetCoefficient(var, 1);
+    constraint->SetBounds(ConstLowBnd, problem->infinity());
```
```suggestion
    auto* constraint = problem->MakeRowConstraint(ConstLowBnd, problem->infinity(), name);
    constraint->SetCoefficient(var, 1);
```

guilpier-code replied:
Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2249#discussion_r1681284477), it applies here as well.

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
+    auto constraint = problem->MakeRowConstraint(constraintName);
+    constraint->SetBounds(0, infinity);
+    return problem;
+}
```
```suggestion
std::tuple<std::unique_ptr<MPSolver>, MPVariable*, MPVariable*> createProblem(const std::string& constraintName)
{
    std::unique_ptr<MPSolver> problem(MPSolver::CreateSolver("GLOP"));
    const double infinity = problem->infinity();
    MPVariable* var1 = problem->MakeNumVar(1, infinity, "var1");
    MPVariable* var2 = problem->MakeNumVar(-infinity, -1, "var2");
    auto constraint = problem->MakeRowConstraint(constraintName);
    constraint->SetBounds(0, infinity);
    return {problem, var1, var2};
}
```
So you don't need `LookupVariableOrNull` afterwards.

guilpier-code replied:
I replaced this code by some other code, simpler, and function **createProblem(...)** is now gone.

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
+    constraint->SetBounds(ConstLowBnd, problem->infinity());
+}
+
+std::unique_ptr<MPSolver> createProblemWith_n_violatedConstraints(const int n)
```
```suggestion
std::unique_ptr<MPSolver> createProblemWith_n_MutuallyIncompatibleConstraints(const int n)
```

guilpier-code replied:
What do you mean ? 
These constraints are unfeasible independently from one another : take only one of them, it's infeasible on its own.

flomnes commented on src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
+    return problem;
+}
+
+void addOneVarConstraintToProblem(MPSolver* problem,
+                                  const std::string& name,
+                                  const double& varLowBnd,
+                                  const double& varUpBnd,
+                                  const double& ConstLowBnd)
```
Don't use const references for `double`

guilpier-code replied:
Corrected

flomnes reviewed (commented):
Please pass POD types by non-const value.

Consider using fixtures for tests.

guilpier-code committed:
Improve infeasible pb test : simplify tests by simplifying infeasible problem construction

guilpier-code commented:
> Consider using fixtures for tests.

I simplified unit tests a bit with last commit, and in the actual state, I can't see any fixture to define : 
For me, fixture : 
- are only related to setting test preparation like building test input data
- are used to avoid duplication

I cant' see no such things in the tests we have here.

guilpier-code committed:
Improve infeasible pb test : give a name to the (only) var of constraints we build for tests

Unlike what was done, the only var in the constraint should a name on its own : a name different from the constraint' name

guilpier-code committed:
Improve infeasible pb test : cleaning due to review

guilpier-code committed:
Improve infeasible pb test : renaming

guilpier-code committed:
[skip ci] Improve infeasible pb test : remove useless include

flomnes committed:
Merge branch 'develop' into fix/improve-infeasible-pb-tests
